src/arcam/fmj/utils.py

```diff
@@ -56,33 +60,69 @@ def get_uniqueid_from_udn(data) -> Optional[str]:
         return None
 
 
-def get_possibly_invalid_xml(data) -> Any:
-    try:
-        return ElementTree.fromstring(data)
-    except ElementTree.ParseError:
-        _LOGGER.info("Device provide corrupt xml, trying with ampersand replacement")
```

Comment:
does the library handle this case?

Reply:
Thanks for reviewing the code.

I'm pretty sure that I didn't handle this case yet but it might be worth discussing whether the XML processing capability is even needed any more.    What I mean by this is that my implementation of `get_uniqueid_from_host()` doesn't need it because it reads the UDN directly out of the `_udn` field in the SSDP response and doesn't need to read the XML to get it.    Worth noting that the `async_upnp_client` parses the `_udn` field out of the standard `USN` header so I think that it is safe to use.

That said, I also provide a function `get_serial_number_from_host()` which does read the XML to get the device serial number.   I included this function because the serial number is immutable whereas the UDN can change after a firmware update.   I was thinking of suggesting this as an alternative for use in the HA integration.   If you do want to use this then I would be happy to look into making it cope with the malformed XML.   If not then perhaps we should just remove it?

As an aside: at the moment, the HA integration uses `get_uniqueid_from_host(session, host)` but in my implementation the `session` argument isn't needed.   At some point you could replace the call with `get_uniqueid(host)`.

Reply:
I had a think about how to handle this and the least nasty solution that I've come up with so far is a custom UpnpRequester that wraps the actual UpnpRequester and maps the response body if need be.   This would also handle any rogue ampersands in any nested XML requests.

Frankly though, having seen this I'd suggest to just drop `get_serial_number_from_host` instead.

```python
class FMJAiohttpSessionRequester(UpnpRequester):
    """Wraps an AiohttpSessionRequester and fixes corrupt XML"""

    def __init__(self, requester: UpnpRequester) -> None:
        self.requester = requester

    async def async_http_request(
        self,
        method: str,
        url: str,
        headers: Optional[Mapping[str, str]] = None,
        body: Optional[str] = None,
    ) -> Tuple[int, Mapping[str, str], str]:

        (
            status_code,
            response_headers,
            response_body,
        ) = await self.requester.async_http_request(method, url, headers, body)

        if status_code == 200:
            try:
                ElementTree.fromstring(response_body)
            except ElementTree.ParseError as err:
                response_body = re.sub(
                    r"&(?![A-Za-z]+[0-9]*;|#[0-9]+;|#x[0-9a-fA-F]+;)",
                    r"&amp;",
                    response_body,
                )

        return status_code, response_headers, response_body


async def get_serial_number_from_host(session: aiohttp.ClientSession, host: str):
    """Get the serial number from a host based on ssdp/upnp."""
    location = await get_upnp_field(host, "location")
    if location is None:
        return None

    requester = FMJAiohttpSessionRequester(AiohttpSessionRequester(session, with_sleep=True))
    factory = UpnpFactory(requester)
    device = await factory.async_create_device(location)

    return device.serial_number  # More immutable than device.udn

```